lib/vrl/stdlib/src/encode_key_value.rs

```diff
@@ -0,0 +1,347 @@
+
+        let key_value_delimiter = arguments
+            .optional("key_value_delimiter")
+            .unwrap_or_else(|| expr!("="));
+
+        let field_delimiter = arguments
+            .optional("field_delimiter")
+            .unwrap_or_else(|| expr!(" "));
+
+        Ok(Box::new(EncodeKeyValueFn {
+            value,
+            fields,
+            key_value_delimiter,
+            field_delimiter,
+        }))
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "encode object",
+                source: r#"encode_key_value!({"lvl": "info", "msg": "This is a message", "log_id": 12345})"#,
+                result: Ok(r#"s'log_id=12345 lvl=info msg="This is a message"'"#),
+            },
+            Example {
+                title: "encode object with fields ordering",
+                source: r#"encode_key_value!({"msg": "This is a message", "lvl": "info", "log_id": 12345}, ["lvl", "msg"])"#,
+                result: Ok(r#"s'lvl=info msg="This is a message" log_id=12345'"#),
+            },
+            Example {
+                title: "custom delimiters",
+                source: r#"encode_key_value!({"start": "ool", "end": "kul", "stop1": "yyc", "stop2" : "gdx"}, key_value_delimiter: ":", field_delimiter: ",")"#,
+                result: Ok(r#"s'end:kul,start:ool,stop1:yyc,stop2:gdx'"#),
+            },
+        ]
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct EncodeKeyValueFn {
+    pub(crate) value: Box<dyn Expression>,
+    pub(crate) fields: Option<Box<dyn Expression>>,
+    pub(crate) key_value_delimiter: Box<dyn Expression>,
+    pub(crate) field_delimiter: Box<dyn Expression>,
+}
+
+fn resolve_fields(fields: Value) -> Result<Vec<String>, ExpressionError> {
+    let arr = fields.try_array()?;
+    arr.iter()
+        .enumerate()
+        .map(|(idx, v)| {
+            v.try_bytes_utf8_lossy()
+                .map(|v| v.to_string())
+                .map_err(|e| format!("invalid field value type at index {}: {}", idx, e).into())
+        })
+        .collect()
+}
+
+impl Expression for EncodeKeyValueFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let fields = match &self.fields {
+            None => Ok(vec![]),
+            Some(expr) => {
+                let fields = expr.resolve(ctx)?;
+                resolve_fields(fields)
+            }
+        }?;
+
+        let object = value.try_object()?;
+
+        let value = self.key_value_delimiter.resolve(ctx)?;
+        let key_value_delimiter = value.try_bytes_utf8_lossy()?;
+
+        let value = self.field_delimiter.resolve(ctx)?;
+        let field_delimiter = value.try_bytes_utf8_lossy()?;
+
+        Ok(encode(object, &fields[..], &key_value_delimiter, &field_delimiter).into())
+    }
+
+    fn type_def(&self, _state: &state::Compiler) -> TypeDef {
+        TypeDef::new().bytes().fallible()
```

Comment:
It seems like we can make this function infallible if the `fields_ordering` argument is omitted?

Reply:
I think so as well, I'll give it a try.